File: src/main/java/de/moritzf/quota/dto/UsageResponseDto.java
package de.moritzf.quota.dto;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;

/**
 * DTO for the top-level usage response payload.
 */
@JsonIgnoreProperties(ignoreUnknown = true)
public record UsageResponseDto(
        @JsonProperty("user_id") String userId,
        @JsonProperty("account_id") String accountId,
        @JsonProperty("email") String email,
        @JsonProperty("rate_limit") RateLimitDto rateLimit,
        @JsonProperty("code_review_rate_limit") RateLimitDto codeReviewRateLimit,
        @JsonProperty("plan_type") String planType
) {
}

File: src/main/kotlin/de/moritzf/quota/dto/UsageResponseDto.kt
package de.moritzf.quota.dto

import de.moritzf.quota.OpenAiCodexQuota
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * DTO for the top-level usage response payload.
 */
@Serializable
data class UsageResponseDto(
    @SerialName("user_id") val userId: String? = null,
    @SerialName("account_id") val accountId: String? = null,
    val email: String? = null,
    @SerialName("rate_limit") val rateLimit: RateLimitDto? = null,
    @SerialName("code_review_rate_limit") val codeReviewRateLimit: RateLimitDto? = null,
    @SerialName("plan_type") val planType: String? = null,
) {
    fun toQuota(): OpenAiCodexQuota {
        return OpenAiCodexQuota(
            primary = rateLimit?.primaryUsageWindow(),
            secondary = rateLimit?.secondaryUsageWindow(),
            reviewPrimary = codeReviewRateLimit?.primaryUsageWindow(),
            reviewSecondary = codeReviewRateLimit?.secondaryUsageWindow(),
            planType = planType.normalizedText(),
            allowed = rateLimit?.allowed,
            limitReached = rateLimit?.limitReached,
            reviewAllowed = codeReviewRateLimit?.allowed,
            reviewLimitReached = codeReviewRateLimit?.limitReached,
            accountId = accountId.normalizedText(),
            email = email.normalizedText(),
        )
    }

    private fun String?.normalizedText(): String? = this?.takeUnless { it.isEmpty() }
}
